### PR TITLE
Memory Optimizations for OSL Training

### PR DESCRIPTION
**Pull request by NikhilNayak-debug**
## Description

This PR implements critical memory optimizations from commit to enable OSL training without OOM errors.

---

## Summary

- Add factorized linear operations to avoid full weight matrix reconstruction during forward passes  
- Optimize gradient projection with memory-efficient `addmm_` operations 
- Enable training on hardware that previously failed with OOM errors  

---

### 1. Factorized Forward Pass (`_factorized_linear`)

- **Before:** Reconstructed full n×n weight matrices during forward pass  
- **After:** Direct factorized computation using SVD components  
- **Impact:** ~40GB memory savings per forward pass on Llama 8B (224 parameters × 192MB each)  

### 2. Optimized Gradient Projection

- Memory-efficient projection: replaced complex matrix operations with `addmm_` operations  
- **FSDP2 compatibility:** maintains distributed training compatibility  

### 3. Updated Forward Pass Logic

- Eliminates weight reconstruction: forward passes now use factorized operations exclusively  
- Preserves API compatibility: no changes to external interfaces  

---

## Test Plan

- Verify memory usage reduction during OSL training
- Ensured FSDP2 compatibility is maintained
- Tested with meta-llama/Llama-3.1-8B-Instruct model

### CODE REVIEW

File: src/mini_trainer/osft_utils.py [hunk 1]
```diff
@@ -10,6 +10,9 @@
 
 from mini_trainer.utils import log_rank_0, check_distributed_is_synchronized
 
+# Memory optimization constants
+OSFT_CACHE_CLEAR_INTERVAL = 5  # Clear GPU cache every N parameters during matrix reconstruction
```
Comment (RobotSail) [inline src/mini_trainer/osft_utils.py:14]: Would it make sense to make this configurable?

Comment (NikhilNayak-debug) [reply src/mini_trainer/osft_utils.py:14]: I don't think this needs to be configurable. This is an internal memory optimization detail during checkpoint saving similar to how we wouldn't expose garbage collection frequency or other memory management params. Value of 5 works but making it configurable would add unnecessary API complexity.

Comment (RobotSail) [reply src/mini_trainer/osft_utils.py:14]: @NikhilNayak-debug It could be worthwhile to have it read this value from an env like:
```suggestion
OSFT_CACHE_CLEAR_INTERVAL = os.getenv('OSFT_CACHE_CLEAR_INTERVAL', 5)
```

So that it still uses the default but is something we can easily adjust if needed.

File: src/mini_trainer/osft_utils.py [hunk 2]
```diff
@@ -900,6 +907,95 @@ def _reconstruct_weight(
...
+            
+            # Ensure all tensors are on the same device and dtype as input
+            device = x.device
+            dtype = x.dtype
+            
+            # Handle both 2D and 3D input tensors
+            original_shape = x.shape
+            if x.dim() == 3:
+                # Flatten 3D input [batch, seq, hidden] -> [batch*seq, hidden] 
+                batch_size, seq_len, hidden_dim = x.shape
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/osft_utils.py:935]: Yeah it should be fine since computation will still take place along the expected dimension

File: src/mini_trainer/osft_utils.py (hunk 1 above)
Comment (NikhilNayak-debug) [reply src/mini_trainer/osft_utils.py]: Sounds good made the change!

File: src/mini_trainer/osft_utils.py [hunk 3]
```diff
@@ -956,6 +1059,19 @@ def prepare_state_dict_for_save(self, state_dict):
                     upcast_dtype=self.upcast_dtype,
                 )
                 state_dict[orig] = W
+                
+                # Explicitly delete intermediate tensors to free memory
+                del U_high, S_high, V_high, U_low, S_low, V_low
```
Comment (RobotSail) [inline src/mini_trainer/osft_utils.py:1064]: @NikhilNayak-debug This doesn't delete the tensors altogether? If we are popping them from the state dict then it would be a pass by reference, have you tested that these are copies and deleting them here doesn't also delete them from the state dict?

**Review by RobotSail — APPROVED** [review]
Great work @NikhilNayak-debug ! Just need some tests and then this is good
